### PR TITLE
lock_manager: make has_waiter accurate

### PR DESCRIPTION
**Pull request by sticnarf**
###  What have you changed?

Now we use an `AtomicBool` variable `has_waiter` to see if there is a waiter so that we can avoid generate hashes when there is no waiter. There is a risk that `has_waiter` is false while there is a waiter, making the waiter not waked up.

This PR uses an `AtomicUsize` to record the waiter count. The increase of this counter is always up to date so there is no such a risk.

###  What is the type of the changes?

Bugfix (a change which fixes an issue)

###  How is the PR tested?

Unit test.

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

###  Benchmark result if necessary (optional)

I don't do an integration benchmark. But from analysis, an operation to an `AtomicU64` is about 10ns, there should be quite minor influence on a single request (which typically needs hundreds of microseconds).

### CODE REVIEW

File: src/server/lock_manager/mod.rs [hunk 1]
```diff
@@ -71,15 +71,7 @@ impl LockManager {
...
-            self.detector_scheduler.clone(),
-            cfg,
-        );
-        self.waiter_mgr_worker
-            .as_mut()
-            .expect("worker should be some")
-            .start(waiter_mgr_runner)?;
+        self.start_waiter_manager(cfg)?;
 
         let detector_runner = Detector::new(
```
Comment (youjiali1995) [inline src/server/lock_manager/mod.rs]: Better to extract a `start/stop_deadlock_detector` to keep consistent.

File: src/server/lock_manager/waiter_manager.rs [hunk 2]
```diff
@@ -101,14 +101,12 @@ impl WaitTable {
 
     fn remove(&mut self, ts: u64) {
         self.wait_table.remove(&ts);
-        if self.wait_table.is_empty() {
-            self.has_waiter.store(false, Ordering::Relaxed);
-        }
+        self.waiter_count.fetch_sub(1, Ordering::SeqCst);
```
Comment (youjiali1995) [inline src/server/lock_manager/waiter_manager.rs]: You should decrease `waiter_count` in `get_ready_waiter()` and `remove_waiter()`.

File: src/server/lock_manager/waiter_manager.rs [hunk 3]
```diff
@@ -488,22 +486,15 @@ mod tests {
...
             .is_some());
-        assert_eq!(has_waiter.load(Ordering::Relaxed), false);
-        wait_table.add_waiter(2, dummy_waiter(1, 2, 2));
-        wait_table.add_waiter(3, dummy_waiter(2, 3, 3));
-        assert_eq!(has_waiter.load(Ordering::Relaxed), true);
-        wait_table.get_ready_waiters(2, vec![2]);
-        assert_eq!(has_waiter.load(Ordering::Relaxed), true);
-        wait_table.get_ready_waiters(3, vec![3]);
-        assert_eq!(has_waiter.load(Ordering::Relaxed), false);
+        assert_eq!(waiter_count.load(Ordering::SeqCst), 0);
```
Comment (youjiali1995) [inline src/server/lock_manager/waiter_manager.rs]: Need to add a test that more than one transactions are waiting for the same transaction.

File: src/server/lock_manager/waiter_manager.rs (hunk 2 above)
Comment (sticnarf) [reply src/server/lock_manager/waiter_manager.rs]: Sorry I misunderstood the original code, will fix.

**Comment by sticnarf** [comment]
@youjiali1995 PTAL again. Thanks!

**Pull request merged**